src/eos-updater-apply.c

```diff
@@ -73,41 +77,506 @@ get_test_osname (void)
+  g_autoptr(GPtrArray) resolved_flatpaks = g_ptr_array_new_with_free_func (g_object_unref);
+  g_autoptr(GHashTable) explicitly_removed_flatpaks = NULL;
+  g_autoptr(GHashTable) already_installed_flatpaks = NULL;
+  gsize i;
+
+  already_installed_flatpaks = determine_set_of_installed_applications (installation, cancellable, error);
+
+  if (!already_installed_flatpaks)
+    return NULL;
+
+  /* Filter the flatpaks this commit wants by removing any flatpaks
+   * which were already explicitly removed earlier, or which are already
+   * installed */
+  for (i = 0; i < flatpaks_this_commit_wants->len; ++i)
+    {
+      FlatpakRef *flatpak_to_install = g_ptr_array_index (flatpaks_this_commit_wants, i);
+
+      if (!g_hash_table_contains (already_installed_flatpaks, flatpak_to_install))
+        {
+          g_ptr_array_add (resolved_flatpaks,
+                           g_object_ref (flatpak_to_install));
+        }
+    }
+
+  return g_steal_pointer (&resolved_flatpaks);
+}
+
+/* Clean up any flatpaks in flatpaks_to_deploy up until right_bound. This
+ * function is called in cases where one of the flatpaks failed to pull
+ * for some reason and we need to roll back the other pulled flatpaks. This
+ * function "never fails", in the sense that if a rollback fails
+ * the best we can do is just warn about it and move on. */
+static void
+cleanup_undeployed_flatpaks_in_installation (FlatpakInstallation *installation,
+                                             GPtrArray           *flatpaks_to_deploy,
+                                             gsize               right_bound)
+{
+  g_autoptr(GError) error = NULL;
+  gsize i = 0;
+
+  g_return_if_fail (right_bound <= flatpaks_to_deploy->len);
+
+  for (; i < right_bound; ++i)
+    {
+      FlatpakRemoteRef *to_deploy = g_ptr_array_index (flatpaks_to_deploy, i);
+      const char *remote_name = flatpak_remote_ref_get_remote_name (to_deploy);
+      g_autofree gchar *ref = flatpak_ref_format_ref (FLATPAK_REF (to_deploy));
+
+      if (!flatpak_installation_remove_local_ref_sync (installation,
+                                                       remote_name,
+                                                       ref,
+                                                       NULL,
+                                                       &error))
+        {
+          g_warning ("Couldn't clean up undeployed ref %s:%s: %s",
+                     remote_name,
+                     ref,
+                     error->message);
+          g_clear_error (&error);
+        }
+    }
+
+  if (!flatpak_installation_prune_local_repo (installation, NULL, &error))
+    {
+      g_warning ("Could not prune orphaned objects in local repo: %s",
+                 error->message);
+      g_clear_error (&error);
+    }
+}
+
+static gboolean
+cleanup_undeployed_flatpaks (GPtrArray *flatpaks_to_deploy,
+                             gsize      right_bound,
```

Comment:
This parameter is set for every call site to `flatpaks_to_deploy->len`, so could be removed from this function.

Reply:
Strictly speaking, that's true, though this function is meant to be a wrapper around `cleanup_undeployed_flatpaks_in_installation` and have a similar API (the latter only exists because there's no point in getting another `FlatpakInstallation` if the caller already has one), so I think for consistency's sake it'd be better to keep it this way.